### ISSUE TITLE
Prototype Azure on ICE

### ISSUE THREAD

**Issue by jekram** [Cloudkibo/ice#11]
Objective: Do prototype with minimal code modification.

1. Fork the AWS ICE code
2. Lay out what is the input diff between AWS and Azure
3. Modify the input
4. Use a dedicated dorplet
5. No need to use Chef for prototype.
6. Feed input via a dummy csv file
7. Standalone no integration with AWS ice.

**Comment by sojharo** [comment]
For this, first I am looking into code again to understand the flow of the code. I need to understand the flow of how the file is located, fetched, downloaded, extracted and parsed. I have started to look into it and once I understand the flow fully, I would be able to download

**Comment by jekram** [comment]
I thought we went thru the code before. We should not wrory about how the file is dowloaded, extracted . The only thing that will be different would be how it is parsed.

**Comment by sojharo** [comment]
Sure, I was just trying to see where it gets the name of the file which we need to download from the bucket. Also, I was looking if there is something more useful that I might get. I had seen the what functions do we have in all the classes, but had not done dry run of these functions.

**Comment by jawaid** [comment]
I think you already did. But the best way to understand somebody else code is through logging trace.

**Comment by sojharo** [comment]
Sure. Let me update more on this soon. I would create a file and put on the aws bucket in a zip file.

**Comment by sojharo** [comment]
I can't push the azure code for this. Looks like, some thing got changed on github in this repository. It asks me to merge the changes on the local. Auto-merge might break many things. I would do manual merge later.

I have a azure sample file on bucket and have new folder for ice azure. Currently, I am having issues on extraction of the zip file downloaded from the bucket.

**Comment by jekram** [comment]
We should fork the code and keep a separate copy for Azue ICE and Google ICE

**Comment by sojharo** [comment]
Yesterday, I had modified the processor code in this. which reads the columns. I changed the names of the columns to the names which are in azure csv file. In this, we need to work on following elements:

1. Downloader which downloads the file and parses it from zip format
2. Processor to break the file into several smaller files
3. Uploader which uploads these smaller files into bucket
4. Reader which should now look into the new uploaded files
5. Reader should now process the new columns

This is high level break up of the tasks in this. It would be better to have a repository for ice azure so that I could upload the modified code on it. I tried forking this repository. It did not allow me to fork it into CloudKibo as it is already there. The other option was to fork it into my Github account.

**Comment by jekram** [comment]
Sojharo asak

Thanks for the update. I will create a repo when I am back atvhome

**Comment by sojharo** [comment]
In this, yesterday, I did some trace to understand how it creates the smaller data files for the reader and upload it on bucket. I have modified the code to mimick them but still I could not understand the purpose of all the files which are created for reader.

Lets have a droplet for this. so that we could upload the code there to run. It doesn't run good on windows here. Also, I could not push to the new repository as it also required me to do the fork in my account and then do a pull request. It is same as I fork the original ice repository.

**Comment by sojharo** [comment]
I have pushed the code in Azure repository and also downloaded it on the live server of our droplet... It saved me time to use the current droplet as it already has grails setup. I ran my modified code and got some problems in making small files. Also, I looked into the small files to understand what they contain and what purpose they have for reader. These files contain calculated values, in our case most of them were zero. I would hopefully be able to upload these files on bucket for our reader by today.

**Comment by jekram** [comment]
Thanks for the update. For the droplet I can delete the current and get a new one. Please let me know

**Comment by jekram** [comment]
Ignore my last comment

**Comment by sojharo** [comment]
I tried solving the problem of small files which are being created by processor. It gave errors because it does calculations and then creates the files and puts results in those files. I opened those small files and they are encrypted. The error is because it does calculations on the basis of columns found in the aws file.

Not all columns are same in aws, google and azure. There are some missing columns or some additional columns in other csv files like google and azure. I have prepared comparison of the columns of all three csv files and on basis of this we can decide what calculations can be made on one and not on other.

This would require me to change the logic of calculations in the LineItemProcessor java class. I would like to show the comparison in today's meeting. I would try to determine what possible calculations we can do with this.

**Comment by sojharo** [comment]
Also, it would require some change in reader and on client side too as we might need to show changed calculations. I am looking into the columns and thinking of what calculations we can have.

**Comment by sojharo** [comment]
I worked more on Google yesterday as I got blocked on Azure because of CSV format. It is not complete csv format and also on daily usage data it does not have column for cost (daily usage section in the file). It had one section for date duration (called statement) in which it showed summary of cost. It is different in format than google and aws. Here is the picture of the csv file for azure:

https://acomdpsstorage.blob.core.windows.net/dpsmedia-prod/azure.microsoft.com/en-us/documentation/articles/billing-understand-your-bill/20150612085908/csvsnapshotportal.png

**Comment by jekram** [comment]
Send meeting coordinates

Please cover last two days
________________________________

**Comment by jekram** [comment]
I looked at the link and it very clear what their billing is looking like. We can discuss this on the call.

**Comment by jekram** [comment]
The key is to lay out the diff between AWS vs. Google vs. Azure.

Lay out what is the input diff between AWS and Azure

**Comment by sojharo** [comment]
Salaam,

Here is the complete update on this issue. Following were the targets:

Objective: Do prototype with minimal code modification.

1. Fork the AWS ICE code

We have the Azure ICE forked in a separate repository where we update it for changes we make.

2. Lay out what is the input diff between AWS and Azure

We have a spreadsheet which does comparison between AWS and Azure input csv file. It also does the same for Google too. From that we know which columns are common and which are not. In the csv file, there were 3 sets of columns in azure, I have included the mixture of set of columns in spreadsheet.

3. Modify the input

We had modified the input csv file to reflect the csv file format of that of Azure. However, the csv file of Azure contains three sections in it. These three sections contain different set of columns. The most important section for ice is daily usage section. This section in Azure do not contain the cost column. The other section which is called statement looks more like summary. It contains the column for cost. But it is not for daily usage. It also had very few rows in it.

4. Use a dedicated dorplet

Currently, we are using the same droplet and it gives the permission denied error. It only runs the original code of aws ice and does not run the code of azure ice.

5. No need to use Chef for prototype.

Not using Chef currently.

6. Feed input via a dummy csv file

We feed input via dummy csv file and we modified the processor in it and processor gives error.  The dummy csv file contains the columns of third section of original csv file.

7. Standalone no integration with AWS ice.

It is completely separate project. The code of aws ice is very specific to the needs of aws and needs modification in code on many places including both server and client side.

I think the prototype for Azure ICE has to be discussed for csv file format. Main issue is the calculation of daily usage cost which can not be calculated by the given csv file.

**Comment by sojharo** [comment]
I have pushed the code for this. I have tried to use the columns of section 2 of csv file, although they are not daily usage columns but they do contain the column for cost. I tried these but this time it is some java run time error on azure. I am going to try this on our droplet if it doesn't give permission error on droplet.

**Comment by sojharo** [comment]
It gives permission denied error on droplet. I am trying to fix java run time error on localhost. It also had happened with aws ice once on local host.

**Comment by jekram** [comment]
Should I get you a separate droplet?

**Comment by sojharo** [comment]
Yes.

On Mon, Jun 22, 2015 at 7:55 AM, Cloudkibo <notifications@github.com> wrote:

> Should I get you a separate droplet?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Cloudkibo/ice/issues/11#issuecomment-113986999>.
>



-- 
Regards,

Sojharo

**Comment by jekram** [comment]
Let's first understand the output files for AWS before proceeding with Google and Azure.

**Comment by sojharo** [comment]
We have looked into the output files of azure and I have detailed the findings in the comment on separate issue. Next, in this is look into LineItemProcessor code and modify the calculations for both ice and azure. Still the code of ice is very much specific to aws as we have the following java classes in tag group which are specific to columns of aws file and store data of columns in form of objects:

1. Account (can be reused)
2. ApplicationGroup 
3. InstanceOs
4. Operation (specific to type of operations aws lists in file)
5. Product (very specific to aws as lists all the products given to us by aws)
6. Region (very specific to region name format like in aws)
7. ResourceGroup
8. Tag
9. TagType
10. UsageType
11. Zone (very specific to zone name format like in aws)

We would need to leave these classes and make classes like them depending on what columns do we have in google and azure csv files.

There are some java classes which are specific to EC2 and S3 services of AWS. We would not use them in google and azure ice.

Next in this, we can work on following classes:

1. BasicLineItemProcessor
2. BillingFileProcessor
3. DataManager in Reader package
4. ReaderConfig

We can understand and modify them to suit  the specific need of google or azure. Depending on change in them, we may also need to change or create new classes in tag group package. Today, I am starting my work on BasicLineItemProcessor.

**Comment by sojharo** [comment]
I looked into both 1. BasicLineItemProcessor and 2. BillingFileProcessor Classes and did logging and tracing. After reading each columns value on a particular row, they create the object for that column.

E.g for the columns zone, region, product and operation, they have the java classes and then they create the objects for each record. They pass them to ReadWriteData class in form of List. The ReadWriteData class is responsible for writing these values and to the files.

The main calculations they did are two: 1. cost by product and 2. usage by product.

It would not be very feasible to convert the current project to ice google or ice azure. It would require us to change many things in many classes on server side. Also, the other classes that we might completely remove and/or replace by new ones is described in my above comment.

**Comment by jekram** [comment]
Is AWS a super set of Google or Azure? or there are functions in Azure and Google that does not exisit in AWS?

**Comment by jekram** [comment]
Begin to work on other items while we decide what to do with ICE for Google and ICE for Azure

**Comment by sojharo** [comment]
Salaam,

AWS is not super set as columns in all three files are different from each other with some similarities. None can be super set of other.

Yes, I have picked other tasks from github too.
